src/Components/Server/src/Circuits/CircuitFactory.cs

```diff
@@ -1,18 +1,114 @@
+            CircuitIdFactory circuitIdFactory,
+            IOptions<CircuitOptions> options)
+        {
+            _scopeFactory = scopeFactory ?? throw new ArgumentNullException(nameof(scopeFactory));
+            _loggerFactory = loggerFactory ?? throw new ArgumentNullException(nameof(loggerFactory));
+            _circuitIdFactory = circuitIdFactory ?? throw new ArgumentNullException(nameof(circuitIdFactory));
+            _options = options?.Value ?? throw new ArgumentNullException(nameof(options));
+            _logger = _loggerFactory.CreateLogger<CircuitFactory>();
+        }
+
+        public CircuitHost CreateCircuitHost(
+            IReadOnlyList<ComponentDescriptor> components,
             CircuitClientProxy client,
             string baseUri,
             string uri,
-            ClaimsPrincipal user);
+            ClaimsPrincipal user)
+        {
+            var scope = _scopeFactory.CreateScope();
+            var jsRuntime = (RemoteJSRuntime)scope.ServiceProvider.GetRequiredService<IJSRuntime>();
+            jsRuntime.Initialize(client);
+
+            var navigationManager = (RemoteNavigationManager)scope.ServiceProvider.GetRequiredService<NavigationManager>();
+            var navigationInterception = (RemoteNavigationInterception)scope.ServiceProvider.GetRequiredService<INavigationInterception>();
+            if (client.Connected)
+            {
+                navigationManager.AttachJsRuntime(jsRuntime);
+                navigationManager.Initialize(baseUri, uri);
+
+                navigationInterception.AttachJSRuntime(jsRuntime);
+            }
+            else
+            {
+                navigationManager.Initialize(baseUri, uri);
+            }
+
+            var renderer = new RemoteRenderer(
+                scope.ServiceProvider,
+                _loggerFactory,
+                _options,
+                client,
+                _loggerFactory.CreateLogger<RemoteRenderer>());
+
+            var circuitHandlers = scope.ServiceProvider.GetServices<CircuitHandler>()
+                .OrderBy(h => h.Order)
+                .ToArray();
+
+            var circuitHost = new CircuitHost(
+                _circuitIdFactory.CreateCircuitId(),
+                scope,
+                _options,
+                client,
+                renderer,
+                components,
+                jsRuntime,
+                circuitHandlers,
+                _loggerFactory.CreateLogger<CircuitHost>());
+            Log.CreatedCircuit(_logger, circuitHost);
+
+            // Initialize per - circuit data that services need
+            (circuitHost.Services.GetRequiredService<ICircuitAccessor>() as DefaultCircuitAccessor).Circuit = circuitHost.Circuit;
+            circuitHost.SetCircuitUser(user);
+
+            return circuitHost;
+        }
+
+        private static class Log
+        {
+            private static readonly Action<ILogger, string, string, Exception> _createdConnectedCircuit =
```

Comment:
Is this logging used?

Reply:
this one is, the one below not. I've updated it.